packages/mui-system/src/Unstable_Grid/gridGenerator.ts

```diff
@@ -202,17 +202,16 @@ export const generateGridStyles = ({ ownerState }: Props): {} => {
           ...(ownerState.disableEqualOverflow && {
             margin: `calc(var(--Grid-rowSpacing) * -1) 0px 0px calc(var(--Grid-columnSpacing) * -1)`,
           }),
-          ...(ownerState.nested
-            ? {
-                padding: `calc(var(--Grid-nested-rowSpacing) / 2) calc(var(--Grid-nested-columnSpacing) / 2)`,
-                ...((ownerState.disableEqualOverflow || ownerState.parentDisableEqualOverflow) && {
-                  padding: `calc(var(--Grid-nested-rowSpacing)) 0px 0px calc(var(--Grid-nested-columnSpacing))`,
-                }),
-              }
-            : {
-                '--Grid-nested-rowSpacing': 'var(--Grid-rowSpacing)',
-                '--Grid-nested-columnSpacing': 'var(--Grid-columnSpacing)',
-              }),
+          ...(ownerState.nested && {
+            padding: `calc(var(--Grid-nested${ownerState.nested}-rowSpacing) / 2) calc(var(--Grid-nested${ownerState.nested}-columnSpacing) / 2)`,
+            ...((ownerState.disableEqualOverflow || ownerState.parentDisableEqualOverflow) && {
+              padding: `calc(var(--Grid-nested${ownerState.nested}-rowSpacing)) 0px 0px calc(var(--Grid-nested${ownerState.nested}-columnSpacing))`,
+            }),
+          }),
+          // variables for the next level grid container
+          [`--Grid-nested${(ownerState.nested ?? 0) + 1}-rowSpacing`]: 'var(--Grid-rowSpacing)',
```

Comment:
Wouldn't it be simpler to add the nested variables even if the grid is nested? That way we shouldn't need any leveling information.

Reply:
Not possible because the variables have 2 purposes:
- consumed by the element itself (margin)
- consumed by the next level grid (padding)

If the variables are the same name, they will behave incorrectly:

```js
// Ex. style of the nested grid container level 1
padding: var(--Grid-rowSpacing); // --Grid-rowSpacing is from the root grid
--Grid-nested-rowSpacing: 16px;
margin: calc(var(--Grid-nested-rowSpacing) * -1);


// style of the nested grid container level 2
padding: var(--Grid-nested-rowSpacing); // --Grid-nested-rowSpacing is from the nested grid container level 1
--Grid-nested-rowSpacing: 24px; // ❌ this will cause the padding to behave incorrectly
margin: calc(var(--Grid-nested-rowSpacing) * -1);
```

This is why the level of information is important to create scope variables for each level.

I am able to simplify the logic by removing nested variables and using only the `--Grid-rowSpacing` and `--Grid-columnSpacing`, however, the leveling information is still necessary.

<img src="https://user-images.githubusercontent.com/18292247/216247587-3135c9c1-beda-40d6-b25e-9c873cb2fe49.png" width="600" />

If you are not use nested grid container, you will get a simple `--Grid-rowSpacing` and `--Grid-columnSpacing` as usual.

Reply:
@mnajdova Here is the [sandbox](https://codesandbox.io/s/billowing-surf-j5d0zu?file=/src/App.tsx) to demonstrate the inheritance.

What I like about the latest change is that the inheritance works with just row and spacing variables. The spacing inheritance can be stopped by providing `spacing` prop to that specific container level which is a correct behavior.

Reply:
In order for what I had in mind to work, we would need to bump the specificity, which is not great. This is what I had in mind:

```diff --git a/packages/mui-system/src/Unstable_Grid/gridGenerator.ts b/packages/mui-system/src/Unstable_Grid/gridGenerator.ts
index 148057c105..98201a717b 100644
--- a/packages/mui-system/src/Unstable_Grid/gridGenerator.ts
+++ b/packages/mui-system/src/Unstable_Grid/gridGenerator.ts
@@ -208,6 +208,10 @@ export const generateGridStyles = ({ ownerState }: Props): {} => {
                 ...((ownerState.disableEqualOverflow || ownerState.parentDisableEqualOverflow) && {
                   padding: `calc(var(--Grid-nested-rowSpacing)) 0px 0px calc(var(--Grid-nested-columnSpacing))`,
                 }),
+                // for nested grids, set the nested variables to the current rowSpacing and columnSpacing if they are defined
+                '& .MuiGrid2-root': {
+                  '--Grid-nested-rowSpacing': 'var(--Grid-rowSpacing, var(--Grid-nested-rowSpacing))',
+                  '--Grid-nested-columnSpacing': 'var(--Grid-columnSpacing, var(--Grid-nested-columnSpacing))',
+                }
               }
             : {
                 '--Grid-nested-rowSpacing': 'var(--Grid-rowSpacing)',
```

I will let to you the decision of what you think it's better. The drawbacks on the two respective solutions are:

- different CSS variable - you will need to know the level in order to redefine it locally
- bumping the specificity, people will need a bigger specificity in order to override it

As we already have a first class API for changing these values - the props of the component, I don't think it matters that the component's variable name differ, so maybe the initially proposed solution is better :)

Reply:
@mnajdova

1. CSS variables does not work if it refers to themselves, so they should be:

    ```diff
    + '& .MuiGrid2-root': {
    +     '--Grid-nested-rowSpacing': 'var(--Grid-rowSpacing)',
    +     '--Grid-nested-columnSpacing': 'var(--Grid-columnSpacing)',
    +  }
    ```

2. What you propose would not work if you specify `spacing` prop to the nested grid.

let's say that this is the style of a deeply nested grid (it could be any level), the padding of the nested grid should use the parent spacing variable to behave correctly:

```js
// assume that --Grid-nested-rowSpacing is 8px;
padding: `calc(var(--Grid-nested-rowSpacing) / 2) calc(var(--Grid-nested-columnSpacing) / 2)`
```

Now if you add the CSS specificity together with spacing prop:

the styles become:

```js
--Grid-nested-rowSpacing: 16px;
--Grid-nested-columnSpacing: 16px;
'& > .MuiGrid2-container': {
  '--Grid-nested-rowSpacing': 'var(--Grid-rowSpacing)',
  '--Grid-nested-columnSpacing': 'var(--Grid-columnSpacing)',
}

// --Grid-nested-rowSpacing resolves to 16px (itself) instead of 8px (parent) ❌ This is wrong.
padding: `calc(var(--Grid-nested-rowSpacing) / 2) calc(var(--Grid-nested-columnSpacing) / 2)`
```

Reply:
My latest change fixes the problem and is also better in one aspect. It does not produce nested variables if you don't have nested grids.

Reply:
> CSS variables does not work if it refers to themselves, so they should be:

Ah I didn't know this, I guess for this to work we need some kind of internal variable as described in the answer here: https://stackoverflow.com/questions/59015022/css-variables-defaults-set-if-not-already-set

I understand now what this wouldn't work, thanks for the explanation.